File: src/js/directable/directables/dotmatrix/shapes/Shape.js






namespace UnityTest.UserInterface.DotMatrix.Shapes
{
    export default class Shape
    {
        DotManager #dotManager;
        List<Vector2Int> #positionGrids = new List<Vector2Int>();
        int #positionGridsIndex = 0;

        bool #isComplete = false;

        // _____________________________________________________________________

        Shape(DotManager dotManager)
        {
            // Store
            #dotManager = dotManager;
        }

        // ______________________________________________________________ Update

        virtual Update()
        {
            // Debug.Log("Shape. Update " + #positionGridsIndex);

            // Complete ?
            if (#isComplete)
            {
                return;
            }

            // Get Dot Index
            int dotIndex = #dotManager.GetNextFreeDotIndex();

            // Clear Current Dot
            #dotManager.ClearDot(dotIndex);

            // if (dotIndex == -1)
            // {
            //     Debug.Log("ShapeLineHorizontal. No more free Dots");
            //     break;
            // }

            // TODO -1 Off Grid ?

            // Position
            #dotManager.SetDotPosition(dotIndex, #positionGrids[#positionGridsIndex]);

            // Fill Dot
            #dotManager.FillDot(dotIndex);

            // Increment Index
            #positionGridsIndex++;

            // Check Complete
            if (#positionGridsIndex >= #positionGrids.Count)
            {
                #isComplete = true;
                return;
            }
        }
    }
}

File: src/js/directable/directables/dotmatrix/Shapes/Shape.cs
using System.Collections.Generic;

using UnityEngine;

using UnityTest.UserInterface.DotMatrix.Dots;

namespace UnityTest.UserInterface.DotMatrix.Shapes
{
    public class Shape
    {
        public DotManager m_dotManager;
        public List<Vector2Int> m_positionGrids = new List<Vector2Int>();
        public int m_positionGridsIndex = 0;

        public bool m_isComplete = false;

        // _____________________________________________________________________

        public Shape(DotManager dotManager)
        {
            // Store
            m_dotManager = dotManager;
        }

        // ______________________________________________________________ Update

        public virtual void Update()
        {
            // Debug.Log("Shape. Update " + m_positionGridsIndex);

            // Complete ?
            if (m_isComplete)
            {
                return;
            }

            // Get Dot Index
            int dotIndex = m_dotManager.GetNextFreeDotIndex();

            // Clear Current Dot
            m_dotManager.ClearDot(dotIndex);

            // if (dotIndex == -1)
            // {
            //     Debug.Log("ShapeLineHorizontal. No more free Dots");
            //     break;
            // }

            // TODO -1 Off Grid ?

            // Position
            m_dotManager.SetDotPosition(dotIndex, m_positionGrids[m_positionGridsIndex]);

            // Fill Dot
            m_dotManager.FillDot(dotIndex);

            // Increment Index
            m_positionGridsIndex++;

            // Check Complete
            if (m_positionGridsIndex >= m_positionGrids.Count)
            {
                m_isComplete = true;
                return;
            }
        }
    }
}